lld/ELF/Relocations.cpp

```diff
@@ -1444,6 +1444,32 @@ template <class ELFT, class RelTy> void RelocationScanner::scanOne(RelTy *&i) {
     }
   }
 
+  if (config->emachine == EM_AARCH64 && type == R_AARCH64_AUTH_ABS64) {
+    // Assume relocations from relocatable objects are RELA.
+    assert(RelTy::IsRela);
+    std::lock_guard<std::mutex> lock(relocMutex);
+    // For a preemptible symbol, we can't use a relative relocation. For an
+    // undefined symbol, we can't compute offset at link-time and use a relative
+    // relocation. Use a symbolic relocation instead.
+    Partition &part = sec->getPartition();
+    if (sym.isPreemptible || sym.isUndefined()) {
+      part.relaDyn->addSymbolReloc(type, *sec, offset, sym, addend, type);
+    } else if (part.relrAuthDyn && sec->addralign >= 2 && offset % 2 == 0 &&
+               isInt<32>(sym.getVA(addend))) {
+      // Implicit addend is below 32-bits so we can use the compressed
+      // relative relocation section. The R_AARCH64_AUTH_RELATIVE
+      // has a smaller addend fielf as bits [63:32] encode the signing-schema.
+      sec->addReloc({expr, type, offset, addend, &sym});
+      part.relrAuthDyn->relocsVec[parallel::getThreadIndex()].push_back(
+          {sec, offset});
+    } else {
+      part.relaDyn->addReloc({R_AARCH64_AUTH_RELATIVE, sec, offset,
+                              DynamicReloc::AddendOnlyWithTargetVA, sym, addend,
+                              R_ABS});
+    }
+    return;
+  }
+
```

Comment:
I need to think about this more, but this does not handle R_AARCH64_AUTH_ABS64 in read-only section (text section).
Any idea this isn't placed near `if (canWrite) {` in `processAux`?

Reply:
Yes, currently we just emit AUTH relocation even against read-only sections with `-z text`, which should not be the case and the error should be emitted (if I'm not mistaken). Thanks for bringing attention to this. I'll prepare a fix for that and submit it shortly. Things seem to be a little bit different from what we already have after `if (canWrite)`: with regular relocs, we emit a plt entry for function symbols, while with auth relocs we probably should just emit an error for all symbol kinds. Anyway, the fix seems easy enough.

Reply:
See fix in 594f8a0e8331b5d11f3efc58fcaa7eae4b9fd7b4. Basically, it disallows auth relocations if `canWrite` is false. Emitting plt entries for `STT_FUNC` symbols and copy relocations for `STT_OBJECT` ones looks meaningless in context of auth relocations. Existing error messages did not seem to make sense, so I've added a new if statement for this particular case with a different message text - not sure if it's applicable though. Please let me know your thoughts on the change.